compiler/rustc_passes/src/liveness.rs

```diff
@@ -318,49 +318,39 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
         NestedVisitorMap::OnlyBodies(self.tcx.hir())
     }
 
-    fn visit_fn(
-        &mut self,
-        fk: FnKind<'tcx>,
-        decl: &'tcx hir::FnDecl<'tcx>,
-        body_id: hir::BodyId,
-        sp: Span,
-        id: HirId,
-    ) {
-        debug!("visit_fn {:?}", id);
+    fn visit_body(&mut self, body: &'tcx hir::Body<'tcx>) {
+        debug!("visit_body {:?}", body.id());
 
-        // swap in a new set of IR maps for this function body:
-        let def_id = self.tcx.hir().local_def_id(id);
-        let mut fn_maps = IrMaps::new(self.tcx);
+        // swap in a new set of IR maps for this body
+        let mut maps = IrMaps::new(self.tcx);
+        let hir_id = maps.tcx.hir().body_owner(body.id());
+        let def_id = maps.tcx.hir().local_def_id(hir_id);
 
         // Don't run unused pass for #[derive()]
-        if let FnKind::Method(..) = fk {
-            let parent = self.tcx.hir().get_parent_item(id);
+        if let Node::ImplItem(..) = self.tcx.hir().get(hir_id) {
+            let parent = self.tcx.hir().get_parent_item(hir_id);
             if let Some(Node::Item(i)) = self.tcx.hir().find(parent) {
                 if i.attrs.iter().any(|a| self.tcx.sess.check_name(a, sym::automatically_derived)) {
                     return;
                 }
             }
         }
```

Comment:
Not sure if the parent relation in the `DefId` space is set up like this, but can you try whether 

```rust
if let Some(parent) = self.tcx.parent(def_id) {
    if self.tcx.has_attr(parent, sym::automatically_derived) {
        return;
    }
}
```

also works? If so, please use that

Reply:
Yep, that works as well.

Part of the motivation for the current form of the check was to avoid using the `automatically_derived` attribute if it were to occur in completely unexpected place, like a crate or a module. Should I change it anyway?

Reply:
I think we should prevent such uses via other means, but you can also add an `assert_eq!(self.tcx.def_kind(parent), DefKind::Impl)` before the return to make sure.

Reply:
We do not prevent such uses this right now, so I used this as a part of condition. I would like to be extra conservative about any additional uses of this attribute since I think it is stabilized unintentionally and its purpose has changed over time.

Reply:
oh :/ that's not great. Yea let's go with a non-panicking version for now